modules/gdnative/pluginscript/register_types.cpp

```diff
@@ -0,0 +1,119 @@
+#include "scene/main/scene_tree.h"
+
+#include "pluginscript_language.h"
+#include "pluginscript_script.h"
+#include <pluginscript/godot_pluginscript.h>
+
+static List<PluginScriptLanguage *> pluginscript_languages;
+
+static Error _check_language_desc(const godot_pluginscript_language_desc *desc) {
+	ERR_FAIL_COND_V(!desc->name || desc->name == String(), ERR_BUG);
+	ERR_FAIL_COND_V(!desc->type || desc->type == String(), ERR_BUG);
+	ERR_FAIL_COND_V(!desc->extension || desc->extension == String(), ERR_BUG);
+	ERR_FAIL_COND_V(!desc->recognized_extensions || !desc->recognized_extensions[0], ERR_BUG);
+	ERR_FAIL_COND_V(!desc->init, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->finish, ERR_BUG);
+
+	// desc->reserved_words is not mandatory
+	// desc->comment_delimiters is not mandatory
+	// desc->string_delimiters is not mandatory
+
+	// desc->get_template_source_code is not mandatory
+	// desc->validate is not mandatory
+
+	// desc->get_template_source_code is not mandatory
+	// desc->validate is not mandatory
+	// desc->find_function is not mandatory
+	// desc->make_function is not mandatory
+	// desc->complete_code is not mandatory
+	// desc->auto_indent_code is not mandatory
+	// desc->add_global_constant is not mandatory
+	// desc->debug_get_error is not mandatory
+	// desc->debug_get_stack_level_count is not mandatory
+	// desc->debug_get_stack_level_line is not mandatory
+	// desc->debug_get_stack_level_function is not mandatory
+	// desc->debug_get_stack_level_source is not mandatory
+	// desc->debug_get_stack_level_locals is not mandatory
+	// desc->debug_get_stack_level_members is not mandatory
+	// desc->debug_get_globals is not mandatory
+	// desc->debug_parse_stack_level_expression is not mandatory
+	// desc->profiling_start is not mandatory
+	// desc->profiling_stop is not mandatory
+	// desc->profiling_get_accumulated_data is not mandatory
+	// desc->profiling_get_frame_data is not mandatory
+	// desc->frame is not mandatory
+
+	ERR_FAIL_COND_V(!desc->script_desc.init, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->script_desc.finish, ERR_BUG);
+
+	ERR_FAIL_COND_V(!desc->script_desc.instance_desc.init, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->script_desc.instance_desc.finish, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->script_desc.instance_desc.set_prop, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->script_desc.instance_desc.get_prop, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->script_desc.instance_desc.call_method, ERR_BUG);
+	ERR_FAIL_COND_V(!desc->script_desc.instance_desc.notification, ERR_BUG);
+	// desc->script_desc.instance_desc.refcount_incremented is not mandatory
+	// desc->script_desc.instance_desc.refcount_decremented is not mandatory
+	return OK;
+}
+
+void GDAPI godot_pluginscript_register_language(const godot_pluginscript_language_desc *language_desc) {
+	Error ret = _check_language_desc(language_desc);
+	if (ret) {
+		ERR_FAIL();
+	}
+	// TODO: @Karroffel how should we pass the langage_desc argument ? Should we only keep a pointer inside PluginScriptLanguage ?
```

Comment:
That's a good question. It depends on if the struct's lifetime is greater than that of the PluginScriptLanguage object aka heap or static data. Should make up a convention for that.

Reply:
My current guess is it's pretty easy for user's library to have a `static` struct of `godot_pluginscript_language_desc` (or store the structure in the `godot_pluginscript_language_data` structure that is currently missing as you pointed out ^^), so we should be able to use the passed pointer during the entire lifetime of the program. The thing is we should add a comment or use a convention to tell the user he should not free this thing until `godot_pluginscript_language_desc.finish` is called

Reply:
This comment can be removed

Reply:
done.